### PR TITLE
Delete flashflood events for bundle tombstones and deletes

### PR DESCRIPTION
**Pull request by xbrianh**

### CODE REVIEW

File: daemons/dss-notify-v2/app.py [hunk 1]
```diff
@@ -16,7 +16,8 @@
 import dss
 from dss import Config, Replica
 from dss.logging import configure_lambda_logging
-from dss.events import get_bundle_metadata_document, get_deleted_bundle_metadata_document, record_event_for_bundle
+from dss.events import (get_bundle_metadata_document, get_deleted_bundle_metadata_document, record_event_for_bundle,
```
Comment (amarjandu) [inline daemons/dss-notify-v2/app.py]: is this supposed to be wrapped in parenthesis ?

File: daemons/dss-notify-v2/app.py [hunk 2]
```diff
@@ -104,23 +105,25 @@ def launch_from_notification_queue(event, context):
...
     def _func(subscription):
         if should_notify(replica, subscription, metadata_document, key):
             notify_or_queue(replica, subscription, metadata_document, key)
 
-    # TODO: Consider scaling parallelization with Lambda size
-    logger.info(f"Attempting notifications for; replica: {replica}, key: {key} delete: {is_delete_event}")
-    with ThreadPoolExecutor(max_workers=20) as e:
+    event_type = metadata_document['event_type']
+    logger.info(f"Attempting notifications for; replica: {replica}, key: {key} event_type: {event_type}")
+    with ThreadPoolExecutor(max_workers=4) as e:
```
Comment (amarjandu) [inline daemons/dss-notify-v2/app.py]: is this reduction in threads intentional?

File: dss/events/__init__.py [hunk 3]
```diff
@@ -54,23 +52,50 @@ def record_event_for_bundle(replica: Replica,
...
+        fqid = key.replace(f".{TOMBSTONE_SUFFIX}", "").split("/")[1]
+        _delete_event_for_bundle(fqid, flashflood_prefixes)
+    elif DSS_BUNDLE_KEY_REGEX.match(key):
+        fqid = key.split("/")[1]
+        _delete_event_for_bundle(fqid, flashflood_prefixes)
+
+def _delete_event_for_bundle(bundle_fqid: str, flashflood_prefixes: typing.Tuple[str, ...]=None):
+    for pfx in flashflood_prefixes:
+        ff = FlashFlood(resources.s3, Config.get_flashflood_bucket(), pfx)
+        ff.delete(event_id=bundle_fqid)
```
Comment (amarjandu) [inline dss/events/__init__.py]: should there be a logger.info call here?

File: tests/test_events.py [hunk 4]
```diff
@@ -41,23 +43,64 @@ def setUp(self):
...
                 self.assertEqual(metadata_document, ret)
                 for args, pfx in zip(ff.call_args_list, used_prefixes):
                     expected = ((resources.s3, Config.get_flashflood_bucket(), pfx),)
                     self.assertEqual(args, expected)
 
+    def test_delete_event_for_bundle(self):
+        keys = [f"bundles/{uuid4()}.{datetime_to_version_format(datetime.utcnow())}",
+                f"bundles/{uuid4()}.{datetime_to_version_format(datetime.utcnow())}.dead",
+                f"bundles/{uuid4()}.dead"]
+        test_parameters = [(replica, key, pfxs)
```
Comment (amarjandu) [inline tests/test_events.py]: could you use itertools.product here?

File: daemons/dss-notify-v2/app.py (hunk 1 above)
Comment (xbrianh) [reply daemons/dss-notify-v2/app.py]: This is an alternative to using `\` to wrap lines

File: daemons/dss-notify-v2/app.py (hunk 2 above)
Comment (xbrianh) [reply daemons/dss-notify-v2/app.py]: Yes, I think 20 was too aggressive.

File: dss/events/__init__.py (hunk 3 above)
Comment (xbrianh) [reply dss/events/__init__.py]: will include

File: tests/test_events.py (hunk 4 above)
Comment (xbrianh) [reply tests/test_events.py]: Possibly, but I'm not sure it would aid readability.

File: daemons/dss-notify-v2/app.py [hunk 5]
```diff
@@ -104,23 +105,25 @@ def launch_from_notification_queue(event, context):
...
         metadata_document = get_deleted_bundle_metadata_document(replica, key)
     else:
-        if exists(replica, key):
-            metadata_document = record_event_for_bundle(replica, key)
-        else:
-            logger.error(f"Key %s not found in replica %s, unable to notify subscribers", key, replica.name)
-            return
+        metadata_document = build_bundle_metadata_document(replica, key)
+    if "CREATE" == metadata_document['event_type']:
+        record_event_for_bundle(replica, key)
```
Comment (chmreid) [inline daemons/dss-notify-v2/app.py]: Should the metadata document that's returned by `record_event_for_bundle()` be saved to `metadata_document` variable? Or is it ignored b/c `record_event_for_bundle()` won't change the metadata document?

File: daemons/dss-notify-v2/app.py [hunk 6]
```diff
@@ -104,23 +105,25 @@ def launch_from_notification_queue(event, context):
...
+def _notify_subscribers(replica: Replica, key: str, metadata_document: dict):
     def _func(subscription):
         if should_notify(replica, subscription, metadata_document, key):
             notify_or_queue(replica, subscription, metadata_document, key)
 
-    # TODO: Consider scaling parallelization with Lambda size
-    logger.info(f"Attempting notifications for; replica: {replica}, key: {key} delete: {is_delete_event}")
-    with ThreadPoolExecutor(max_workers=20) as e:
+    event_type = metadata_document['event_type']
+    logger.info(f"Attempting notifications for; replica: {replica}, key: {key} event_type: {event_type}")
```
Comment (chmreid) [inline daemons/dss-notify-v2/app.py]:
```suggestion
    logger.info(f"Attempting notifications for an event: replica: {replica}, key: {key}, event_type: {event_type}")
```

File: dss/events/__init__.py [hunk 7]
```diff
@@ -3,6 +3,7 @@
 import logging
 import threading
 import typing
+import logging
```
Comment (chmreid) [inline dss/events/__init__.py]: Delete (this is already imported on line 3)

File: dss/events/__init__.py [hunk 8]
```diff
@@ -54,23 +54,51 @@ def record_event_for_bundle(replica: Replica,
...
         event_date = datetime.utcnow()
+    metadata_document = build_bundle_metadata_document(replica, key)
     for pfx in flashflood_prefixes:
         ff = FlashFlood(resources.s3, Config.get_flashflood_bucket(), pfx)
         if not ff.event_exists(fqid):
             ff.put(json.dumps(metadata_document).encode("utf-8"), event_id=fqid, date=event_date)
     return metadata_document
 
-def _build_bundle_metadata_document(replica: Replica, key: str) -> dict:
+def delete_event_for_bundle(replica: Replica, key: str, flashflood_prefixes: typing.Tuple[str, ...]=None):
```
Comment (chmreid) [inline dss/events/__init__.py]: Might be nice to have a comment or two on what cases each of these conditional blocks are handling

File: tests/test_events.py [hunk 9]
```diff
@@ -50,23 +52,64 @@ def setUp(self):
...
-        with mock.patch("dss.events._build_bundle_metadata_document", return_value=metadata_document):
+        keys = [f"bundles/{uuid4()}.{datetime_to_version_format(datetime.utcnow())}",
+                f"bundles/{uuid4()}.{datetime_to_version_format(datetime.utcnow())}.dead",
+                f"bundles/{uuid4()}.dead"]
+        test_parameters = [(replica, key, pfxs)
+                           for replica in Replica
+                           for key in keys
+                           for pfxs in [None, ("foo", "bar")]]
+        for replica, key, prefixes in test_parameters:
+            with self.subTest(replica=replica.name, key=key, flashflood_prefixes=prefixes):
```
Comment (chmreid) [inline tests/test_events.py]: Huh, didn't know you could pass kwargs only to a subTest name. How does this subTest name look when displayed on the screen?

**Pull request closed without merging**